theano/gpuarray/subtensor.py

```diff
@@ -800,7 +800,7 @@ def make_node(self, x, y, ilist):
         """
         ctx_name = infer_context_name(x, y, ilist)
         x_ = as_gpuarray_variable(x, ctx_name)
-        y_ = as_gpuarray_variable(y, ctx_name)
+        y_ = as_gpuarray_variable(y.astype(x.dtype), ctx_name)
```

Comment:
I would prefer not to multiply the possible C code b y including a cartesian product of all the types.

Also, if we do it this way it gives the opportunity to merge the cast with another elemwise operation.

Finally since we have float16 support, there would need to be special-case handling of casting to and from that type which would complicate the kernel code.